# Probability and Information Theory



## Probability Mass Function (PMF) 


> <img src="images/3.Probability/Dice.png" height="30px" width="60px" class="hint" style="text-align:left"/> The domain of X contains discrete values

* The domain of P must be the set of all possible states of x.
 ∀x ∈ x,0≤ P(x)≤1.
* An impossible event has probability 0, and no statecan be less probable than that. Likewise, an event that is guaranteed tohappen has probability 1, and no state can have a greater chance of occurring.•
* x∈xP(x) = 1. We refer to this property as beingnormalized. 

![PMF](images/PMF.png)


Example : Uniformed distribution

$$ P(X = x_i) = \frac{1}{k} \tag{3.1} $$

![Uniformed Discrete](images\3.Probability\UniformDiscrete.png)

## Probability Density Function (PDF)


 * The domain of p must be the set of all possible states of x.
 * ∀x ∈ x, p(x) ≥ 0. Note that we do not require p(x) ≤ 1.
 * $\int p(x)dx = 1 $

Example: Normal Distribution

![Normal Distribution](images/3.Probability/PDF.png)

## Joint, Marginal and Conditional Probability

P(X = x,Y = y) is the joint distribution of X and Y, while P(X = x|Y = y) is the conditional distribution of X given Y
$$
P(X=x) = \Sigma_y P(X=x,Y=y) = \Sigma_y (X=x | Y = y)P(Y=y)
$$


$$
p(x) = \int p(x,y)dy
$$

Following image illustrate the three proability distribution via a two-way table. Highly recommend see [this great video](https://www.youtube.com/watch?v=SrEmzdOT65s) from zedstatistics .

![Probability Distribution](Images/3.Probability/ProbabilityDistribution.png)



In [38]:
import pandas as pd
import numpy as np

tv = pd.read_csv("TVShowProbability.csv",index_col=0) 
tv

Male  Female
Game of Thrones    80     120
West World        100      25
Other              50     125

In [39]:
# Create Marginal Total (column and row)
tv["Total"] = tv.sum(axis=1)
total_row = pd.Series(tv.sum(axis=0))
total_row.name = "RowTotal"
tv = tv.append(total_row)
tv

Male  Female  Total
Game of Thrones    80     120    200
West World        100      25    125
Other              50     125    175
RowTotal          230     270    500

In [40]:
# Convert to two-way frequency probablity table
# P(Gender) = Sum of P(TVShow,Gender) over TVShow
# P(TVShow) = Sum of P(TVShow,Gender) over Gender
tv_prob = tv/tv.ix["RowTotal","Total"]
tv_prob

Male  Female  Total
Game of Thrones  0.16    0.24   0.40
West World       0.20    0.05   0.25
Other            0.10    0.25   0.35
RowTotal         0.46    0.54   1.00

In [41]:
# Add Conditional column over "Female" 
# P(TVShow | Gender) = P(TVShow,Gender) / P(Gender)
tv_prob["Condition | Female"] = tv_prob["Female"] / tv_prob.ix["RowTotal","Female"]
tv_prob

Male  Female  Total  Condition | Female
Game of Thrones  0.16    0.24   0.40            0.444444
West World       0.20    0.05   0.25            0.092593
Other            0.10    0.25   0.35            0.462963
RowTotal         0.46    0.54   1.00            1.000000

## Probablity Calculation



### Chain Rule

> Think joint probability as logic AND operation

$$ P(x^{(1)},...,x^{(n)}) = P(x^{(1)})\prod_{i=2}^{n} P(x^{(i}|x^{(1)},...,x^{(i-1})) $$

$$ P(a,b,c) = P(a|b,c)P(b,c) = P(a|b,c)P(b|c)P(c) $$



In [42]:
# P1(Game of Thrones, Female) = P2 (Game of Thrones | Female) P3(Female)
P1 = tv_prob.ix["Game of Thrones","Female"]
P2 = tv_prob.ix["Game of Thrones", "Condition | Female"]
P3 = tv_prob.ix["RowTotal", "Female"]
print(P1, P2*P3)

0.24 0.24


### Independence

$$\forall x \in X, y \in Y, p(X=x,Y=y) = p(X=x)P(Y=y) $$

$$ X \bot Y $$

In [43]:
# P1(Game of Thrones, Female)
# P2(Game of Thrones)
# P3(Female)
P1 = tv_prob.ix["Game of Thrones","Female"]
P2 = tv_prob.ix["Game of Thrones", "Total"]
P3 = tv_prob.ix["RowTotal", "Female"]
print(P1, P2*P3)
# P1 <> P2* P3, so TVShow and Gender are not independent

0.24 0.216


## Expectation

$$ \mathbb{E}_{x~P}[f(x)]=\Sigma_x P(x)f(x) \tag{3.9} $$

$$ \mathbb{E}_{x~P}[f(x)]=\int_x P(x)f(x)dx \tag{3.10} $$


$$ \mathbb{E}_{x}[\alpha f(x)+\beta g(x)]=\alpha\mathbb{E}_{x}[f(x)] + \beta\mathbb{E}_{x}[g(x)] \tag{3.11} $$

In uniformed distribution, $\mathbb{E}$ become the mean of X.

$$ \mathbb{E}(x) = \frac{1}{N}\sum\limits_i^{n} x_i $$

In [44]:
from sklearn import datasets
boston = datasets.load_boston()
boston_df = pd.DataFrame(boston.data)
boston_df.columns = boston.feature_names
boston_df["PRICE"] = boston.target
boston_df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  PRICE  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

In [45]:
# Expectation of House Price
E_Price = boston_df["PRICE"].mean()
print(E_Price)

22.532806324110698


### Variance and Covariance

$$ \mu = \mathbb{E}(x) $$

$$ Var(x) = (x - \mathbb{E}(x))^2 $$

> Variance measures how far a set of (random) numbers are spread out (Scattered) from their expected value

$$ Cov(f(x),g(y)) = \mathbb{E}\bigg((f(x) - \mathbb{E}(x))(g(y)- \mathbb{E}(g(y)) \bigg)  $$

In uniform distribution, the coveriance becomes below
$$ Cov(X, Y) = \frac{1}{{N}} \sum\limits_{i}^{N}{(X_i – \mu_x)(Y_i – \mu_y)} $$

> Covariance is a measure how **linearly related** two variables are

<a href="http://stats.seandolinar.com/covariance-different-ways-to-explain/"><img src="images/3.Probability/Covariance.png" width="300px" /></a>


**Covariance Matrix**

$$ Cov(x)_{i,j} = Cov(x_i,y_j) $$

$$ Cov(x,x) = Var(x) $$




In [46]:
cov = boston_df.cov()
cov

CRIM           ZN       INDUS      CHAS        NOX         RM  \
CRIM      73.904671   -39.990934   23.854453 -0.120739   0.415924  -1.328491   
ZN       -39.990934   543.936814  -85.412648 -0.252925  -1.396148   5.112513   
INDUS     23.854453   -85.412648   47.064442  0.109669   0.607074  -1.887957   
CHAS      -0.120739    -0.252925    0.109669  0.064513   0.002684   0.016285   
NOX        0.415924    -1.396148    0.607074  0.002684   0.013428  -0.024603   
RM        -1.328491     5.112513   -1.887957  0.016285  -0.024603   0.493671   
AGE       84.886013  -373.901548  124.513903  0.618571   2.385927  -4.751929   
DIS       -6.840937    32.629304  -10.228097 -0.053043  -0.187696   0.303663   
RAD       46.561611   -63.348695   35.549971 -0.016296   0.616929  -1.283815   
TAX      839.717323 -1236.453735  833.360290 -1.523367  13.046286 -34.583448   
PTRATIO    5.364787   -19.776571    5.692104 -0.066819   0.047397  -0.540763   
B       -296.171637   373.721402 -223.579756  1.131325  -4.020570   8.215006   
LSTAT     27.761872   -68.783037   29.580270 -0.097816   0.488946  -3.079741   
PRICE    -30.505977    77.315176  -30.520823  0.409409  -0.455412   4.493446   

                 AGE         DIS          RAD           TAX     PTRATIO  \
CRIM       84.886013   -6.840937    46.561611    839.717323    5.364787   
ZN       -373.901548   32.629304   -63.348695  -1236.453735  -19.776571   
INDUS     124.513903  -10.228097    35.549971    833.360290    5.692104   
CHAS        0.618571   -0.053043    -0.016296     -1.523367   -0.066819   
NOX         2.385927   -0.187696     0.616929     13.046286    0.047397   
RM         -4.751929    0.303663    -1.283815    -34.583448   -0.540763   
AGE       792.358399  -44.329379   111.770846   2402.690122   15.936921   
DIS       -44.329379    4.434015    -9.068252   -189.664592   -1.059775   
RAD       111.770846   -9.068252    75.816366   1335.756577    8.760716   
TAX      2402.690122 -189.664592  1335.756577  28404.759488  168.153141   
PTRATIO    15.936921   -1.059775     8.760716    168.153141    4.686989   
B        -702.940328   56.040356  -353.276219  -6797.911215  -35.059527   
LSTAT     121.077725   -7.473329    30.385442    654.714520    5.782729   
PRICE     -97.589017    4.840229   -30.561228   -726.255716  -10.110657   

                   B       LSTAT       PRICE  
CRIM     -296.171637   27.761872  -30.505977  
ZN        373.721402  -68.783037   77.315176  
INDUS    -223.579756   29.580270  -30.520823  
CHAS        1.131325   -0.097816    0.409409  
NOX        -4.020570    0.488946   -0.455412  
RM          8.215006   -3.079741    4.493446  
AGE      -702.940328  121.077725  -97.589017  
DIS        56.040356   -7.473329    4.840229  
RAD      -353.276219   30.385442  -30.561228  
TAX     -6797.911215  654.714520 -726.255716  
PTRATIO   -35.059527    5.782729  -10.110657  
B        8334.752263 -238.667516  279.989834  
LSTAT    -238.667516   50.994760  -48.447538  
PRICE     279.989834  -48.447538   84.586724

![Covariance Matrix](images/3.probability/CovarianceMatrix.jpeg)
 

In [47]:
boston_df_desc = boston_df.describe()
boston_df_desc

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.593761   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.596783   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.647423   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT       PRICE  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [48]:
# we can see that the square of σ(standard deviation) is equal to Cov(x,x) and Var(x)
boston_df_desc.ix["std","CRIM"]**2, cov.ix["CRIM","CRIM"]

(73.904670956866809, 73.904670956866809)

In [49]:
# get expectation of boston dataframe features
boston_df_E = boston_df.mean(axis=0)
print(boston_df_E)

CRIM         3.593761
ZN          11.363636
INDUS       11.136779
CHAS         0.069170
NOX          0.554695
RM           6.284634
AGE         68.574901
DIS          3.795043
RAD          9.549407
TAX        408.237154
PTRATIO     18.455534
B          356.674032
LSTAT       12.653063
PRICE       22.532806
dtype: float64


In [50]:
# set (x - E(x)) to X_hat
boston_df_hat = boston_df - boston_df_E
boston_df_hat.transpose().dot(boston_df_hat)/boston_df_hat.shape[0]

CRIM           ZN       INDUS      CHAS        NOX         RM  \
CRIM      73.758614   -39.911900   23.807310 -0.120500   0.415103  -1.325865   
ZN       -39.911900   542.861840  -85.243848 -0.252425  -1.393389   5.102410   
INDUS     23.807310   -85.243848   46.971430  0.109452   0.605874  -1.884225   
CHAS      -0.120500    -0.252425    0.109452  0.064385   0.002679   0.016253   
NOX        0.415103    -1.393389    0.605874  0.002679   0.013401  -0.024555   
RM        -1.325865     5.102410   -1.884225  0.016253  -0.024555   0.492695   
AGE       84.718255  -373.162612  124.267828  0.617349   2.381212  -4.742538   
DIS       -6.827417    32.564819  -10.207884 -0.052938  -0.187325   0.303063   
RAD       46.469592   -63.223500   35.479714 -0.016263   0.615710  -1.281277   
TAX      838.057803 -1234.010151  831.713333 -1.520357  13.020502 -34.515101   
PTRATIO    5.354185   -19.737487    5.680855 -0.066687   0.047304  -0.539695   
B       -295.586318   372.982822 -223.137898  1.129089  -4.012624   8.198771   
LSTAT     27.707007   -68.647102   29.521811 -0.097623   0.487980  -3.073655   
PRICE    -30.445689    77.162379  -30.460505  0.408600  -0.454512   4.484566   

                 AGE         DIS          RAD           TAX     PTRATIO  \
CRIM       84.718255   -6.827417    46.469592    838.057803    5.354185   
ZN       -373.162612   32.564819   -63.223500  -1234.010151  -19.737487   
INDUS     124.267828  -10.207884    35.479714    831.713333    5.680855   
CHAS        0.617349   -0.052938    -0.016263     -1.520357   -0.066687   
NOX         2.381212   -0.187325     0.615710     13.020502    0.047304   
RM         -4.742538    0.303063    -1.281277    -34.515101   -0.539695   
AGE       790.792473  -44.241772   111.549955   2397.941723   15.905425   
DIS       -44.241772    4.425252    -9.050331   -189.289761   -1.057680   
RAD       111.549955   -9.050331    75.666531   1333.116741    8.743402   
TAX      2397.941723 -189.289761  1333.116741  28348.623600  167.820822   
PTRATIO    15.905425   -1.057680     8.743402    167.820822    4.677726   
B        -701.551118   55.929604  -352.578045  -6784.476608  -34.990240   
LSTAT     120.838441   -7.458560    30.325392    653.420617    5.771300   
PRICE     -97.396153    4.830663   -30.500830   -724.820428  -10.090676   

                   B       LSTAT       PRICE  
CRIM     -295.586318   27.707007  -30.445689  
ZN        372.982822  -68.647102   77.162379  
INDUS    -223.137898   29.521811  -30.460505  
CHAS        1.129089   -0.097623    0.408600  
NOX        -4.012624    0.487980   -0.454512  
RM          8.198771   -3.073655    4.484566  
AGE      -701.551118  120.838441  -97.396153  
DIS        55.929604   -7.458560    4.830663  
RAD      -352.578045   30.325392  -30.500830  
TAX     -6784.476608  653.420617 -724.820428  
PTRATIO   -34.990240    5.771300  -10.090676  
B        8318.280421 -238.195841  279.436494  
LSTAT    -238.195841   50.893979  -48.351792  
PRICE     279.436494  -48.351792   84.419556

## Common Distribution

### Bernoulli Distribution

$$ P(x=1) = \phi \tag{3.16} $$
$$ P(x=0) = 1 - \phi \tag{3.17} $$
$$ P(x=x) = \phi^x (1-\phi)^{1-x} \tag{3.18} $$
$$ \mathbb{E}[x] = \phi \tag{3.19} $$
$$ Var(x) = \phi(1-\phi) \tag{3.20} $$




## Gaussian Distribution

$$  \mathcal{N}(x;\mu,\sigma^2) = \sqrt{\frac{1}{2\pi\sigma^2}}exp\big(-\frac{1}{2\sigma^2}(x-\mu)^2\big) \tag{3.21} $$

In [51]:
mu, sigma = 0, 0.5

sample = np.random.normal(mu,sigma,10000)
hist, edges = np.histogram(sample, density=True, bins=1000)

x = np.linspace(-2,2,1000)
pdf = 1 / np.sqrt(2*np.pi*(sigma**2)) * np.exp(-1/(2*(sigma**2))*((x-mu)**2))

x[0:2], pdf[:2]

(array([-2.      , -1.995996]), array([ 0.00026766,  0.00027636]))

In [52]:
from bokeh.layouts import gridplot
from bokeh.plotting import figure, show, output_file
from bokeh.io import output_notebook,push_notebook
output_notebook() # output to jupyter notebook

p1 = figure(title="Normal Distribution (μ=0, σ=0.5)")

# histogram
p1.quad(top=hist,bottom=0,left=edges[:-1],right=edges[1:])

# bell curve
p1.line(x,pdf,line_color="#D95B43", line_width=4, alpha=0.7, legend="PDF") 

# sigma line
p1.line([-sigma,-sigma],[0,1.2],line_color="#333333", line_width=2, alpha=0.7,legend="σ")
p1.line([sigma,sigma],[0,1.2],line_color="#333333", line_width=2, alpha=0.7,legend="σ")
p1.line([mu,mu],[0,1.2],line_color="#DDFFAA", line_width=2, alpha=0.7,legend="μ")
show(p1, notebook_handle=True)

def update(mu=0,sigma=0.5): 
    #r.data_source.data['y'] = A * func(w * x + phi)
    push_notebook()


Loading BokehJS ...

In [53]:
from ipywidgets import interact
interact(update, mu=(-1,1,0.1), sigma=(-5,5,0.1))

<function __main__.update>